File: backend/app/models/rattrapage.py
from sqlalchemy import Column, Integer, ForeignKey, Date, Time, DateTime, Enum as SAEnum
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
from app.core.database import Base
from app.models.enums import StatutRattrapage

class Rattrapage(Base):
    __tablename__ = "rattrapages"

    id = Column(Integer, primary_key=True, index=True)
    absence_id = Column(Integer, ForeignKey("absences.id", ondelete="CASCADE"), nullable=False)
    salle_id = Column(Integer, ForeignKey("salles.id", ondelete="CASCADE"), nullable=False)
    date_proposee = Column(Date, nullable=False)
    heure_debut = Column(Time, nullable=False)
    heure_fin = Column(Time, nullable=False)
    statut = Column(SAEnum(StatutRattrapage), default=StatutRattrapage.PROPOSE, nullable=False)
    valide_par = Column(Integer, ForeignKey("utilisateurs.id", ondelete="SET NULL"), nullable=True)
    
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())

    # Relationships
    absence = relationship("Absence", back_populates="rattrapages")
    salle = relationship("Salle", back_populates="rattrapages")
    validateur = relationship("Utilisateur", back_populates="rattrapages_valides")
    emplois_du_temps = relationship("EmploiDuTemps", back_populates="rattrapage", cascade="all, delete-orphan")

File: frontend/src/types/rattrapage.ts
import { Salle } from "./salle";

export type StatutRattrapage = "propose" | "valide" | "annule";

export interface Rattrapage {
  id: number;
  absence_id: number;
  salle_id: number;
  date_proposee: string;
  heure_debut: string;
  heure_fin: string;
  statut: StatutRattrapage;
  valide_par?: number | null;
  created_at: string;
  updated_at: string;
}

export interface RattrapageResponse extends Rattrapage {
  absence?: {
    id: number;
    enseignant_id: number;
    matiere_id: number;
    date_absence: string;
    motif: string;
    justificatif: string | null;
    statut: string;
    enseignant?: {
      id: number;
      nom: string;
      prenom: string;
      email: string;
    };
    matiere?: {
      id: number;
      nom: string;
      departement_id: number;
    };
  };
  salle?: Salle;
  validateur?: {
    id: number;
    nom: string;
    prenom: string;
    email: string;
  } | null;
}

export interface CreateRattrapagePayload {
  absence_id: number;
  salle_id: number;
  date_proposee: string;
  heure_debut: string;
  heure_fin: string;
}

export interface UpdateSallePayload {
  salle_id: number;
}
